pkg/server/server_sql.go

```diff
@@ -600,26 +602,39 @@ func newSQLServer(ctx context.Context, cfg sqlServerArgs) (*SQLServer, error) {
 		cfg.db,
 	)
 
-	// We can't use the nodeDialer as the sqlInstanceDialer unless we
-	// are serving the system tenant despite the fact that we've
-	// arranged for pod IDs and instance IDs to match since the
-	// secondary tenant gRPC servers currently live on a different
-	// port.
-	canUseNodeDialerAsSQLInstanceDialer := isMixedSQLAndKVNode && codec.ForSystemTenant()
-	if canUseNodeDialerAsSQLInstanceDialer {
-		cfg.sqlInstanceDialer = cfg.kvNodeDialer
-	} else {
-		// In a multi-tenant environment, use the sqlInstanceReader to resolve
-		// SQL pod addresses.
-		addressResolver := func(nodeID roachpb.NodeID) (net.Addr, roachpb.Locality, error) {
+	addressResolver := func(nodeID roachpb.NodeID) (net.Addr, roachpb.Locality, error) {
+		if cfg.Settings.Version.IsActive(ctx, clusterversion.V25_3) {
 			info, err := cfg.sqlInstanceReader.GetInstance(cfg.rpcContext.MasterCtx, base.SQLInstanceID(nodeID))
 			if err != nil {
 				return nil, roachpb.Locality{}, errors.Wrapf(err, "unable to look up descriptor for n%d", nodeID)
 			}
-			return &util.UnresolvedAddr{AddressField: info.InstanceRPCAddr}, info.Locality, nil
+			defaultAddress := &util.UnresolvedAddr{AddressField: info.InstanceRPCAddr}
+			return cfg.Locality.LookupAddress(info.LocalityAddressList, defaultAddress), info.Locality, nil
```

Comment:
Could we unify this with the sql instance resolve we use below? If LocalityAddressList is empty this function is a no-op so we should be able to use the same implementation for both in either case.  I understand the desire to keep the old code path unchanged, but that is really only valuable if we also have a cluster setting that lets you opt-opt of it.